triton-vm/src/stark.rs

```diff
@@ -200,21 +192,51 @@ impl Stark {
             ext_codeword_tables.get_all_quotient_degree_bounds(self.num_trace_randomizers);
         timer.elapsed("Calculated quotient degree bounds");
 
-        // Prove equal initial values for the column pairs pertaining to cross table arguments
-        for arg in AllCrossTableArgs::default().into_iter() {
-            quotient_codewords.push(arg.initial_quotient(&ext_codeword_tables, &self.xfri.domain));
-            quotient_degree_bounds
-                .push(arg.quotient_degree_bound(&ext_codeword_tables, self.num_trace_randomizers));
-        }
+        // Prove equal terminal values for the column pairs pertaining to cross table arguments
+        let grand_cross_table_arg_weights_seed = proof_stream.prover_fiat_shamir();
+        let grand_cross_table_arg_weights = Self::sample_weights(
+            grand_cross_table_arg_weights_seed,
+            &hasher,
+            NUM_CROSS_TABLE_ARGS + NUM_PUBLIC_EVAL_ARGS,
+        )
+        .try_into()
+        .expect("Sample exactly as many weights as needed.");
+        let input_terminal = EvalArg::compute_terminal(
+            &self.input_symbols,
+            EvalArg::default_initial(),
+            extension_challenges
+                .processor_table_challenges
+                .input_table_eval_row_weight,
+        );
+        let output_terminal = EvalArg::compute_terminal(
+            &self.output_symbols,
+            EvalArg::default_initial(),
+            extension_challenges
+                .processor_table_challenges
+                .output_table_eval_row_weight,
+        );
+        let grand_cross_table_arg = GrandCrossTableArg::new(
+            &grand_cross_table_arg_weights,
+            input_terminal,
+            output_terminal,
+        );
+        let gcta_quotient_codeword = grand_cross_table_arg.terminal_quotient_codeword(
+            &ext_codeword_tables,
+            &self.xfri.domain,
+            derive_omicron(ext_codeword_tables.padded_height as u64),
+        );
+        quotient_codewords.push(gcta_quotient_codeword);
 
-        let non_lin_combi_weights_seed = proof_stream.prover_fiat_shamir();
-        timer.elapsed("Fiat-Shamir to get seed for sampling non-linear-combination weights");
+        let gcta_quotient_degree_bound = grand_cross_table_arg
+            .quotient_degree_bound(&ext_codeword_tables, self.num_trace_randomizers);
+        quotient_degree_bounds.push(gcta_quotient_degree_bound);
+        timer.elapsed("Grand Cross Table Argument");
 
+        let non_lin_combi_weights_seed = hasher.hash_sequence(&grand_cross_table_arg_weights_seed);
```

Comment:
I'm not super happy with this part of the weight derivation. The main reason I derive the second seed from the first is due to the way the number of to-be-sampled weights is determined. Concretely, the number of weights for the non-linear combination depends on the number of quotients, and the first seed for the grand cross-table argument changes the number of quotients.

Ideally, we know the number of quotients from the start, and sample the required number of weights, instead of deriving one seed from the other. However, hardcoding the number of quotients did not seem like the nice thing to do.

Reply:
I think a possible way forward is to conceptually keep the seed-derivation as-is for now (but change it to use `hash_pair`, as you suggest), and then re-work this part once the number of quotients is stable, i.e., just before we start implementing the recufier.

Reply:
Looking at it again, the “unknown” number of quotients is the number of quotients produced by the grand cross-table argument, which is 1, thus easy to predict. This part should be cleaner now.